### PR TITLE
Add support for locking the registry.

### LINKED ISSUE #1259: [GUI] Refresh button can cause issues with the registry if another process installs a mod
It's possible to have CKAN install a module and have it become "orphaned" if an unrelated GUI was launched before the installation and the refresh button is pressed.

2 repro methods:

Have a GUI instance open, install a mod via CLI, refresh GUI. The entries in the registry created by CLI will be lost and the installed mod(s) will be "orphaned".

Have 2 GUI instances open: `A` and `B`. Install a mod in instance `A`, once install is complete refresh instance `B`. The entries in the registry created by `A` will be lost and the installed mod(s) will be "orphaned".

### PR DESCRIPTION
**Pull request by mgsdk**
This change creates a file when the RegistryManager loads a registry. Any other attempts at opening the registry will throw a kraken. Should be implemented in both cmd and gui.

Closes #1259 and #706.

### CODE REVIEW

File: Core/Registry/RegistryManager.cs [hunk 1]
```diff
@@ -49,6 +60,40 @@ private RegistryManager(string path, KSP ksp)
...
+            ReleaseLock();
+        }
+
+        /// <summary>
+        /// Tries to lock the registry by creating a lock file.
+        /// </summary>
+        /// <returns><c>true</c>, if lock was gotten, <c>false</c> otherwise.</returns>
+        public bool GetLock()
+        {
+            if (File.Exists(lock_file_path))
```
Comment (pjf) [inline Core/Registry/RegistryManager.cs]: Oooh! This looks like a race condition to me!

* Process A checks to see if the lockfile exists, and discovers it doesn't.
* Process B creates the lockfile
* Process A calls `File.Create`, but it's already there!  Bam!

Further discussion regarding locking incoming on the main PR thread. :)

**Comment by pjf** [comment]
It looks like we're using lockfiles; while that's not a terrible solution, it means we an leave them lying around by accident (which is a pain to users) and there's a chance they can get deleted even when our registry is in use (which is disastrous for us).

On a Unix system, we'd use the `flock()` system call, which allows us to lock an open file. That guarantees that nothing else can take the lock away from us, and means that when our process exits (or even closes the file) the lock is released.

The equivalent under .NET seems to be the [fileshare permission](https://msdn.microsoft.com/en-us/library/system.io.fileshare\(v=vs.110\).aspx), which allows the `FileShare.None` property be expressed, which prevents any other processes (including threads in the same process) from accessing the file. One possible solution would then be:

- Mark the registry as `:IDisposable` if it's not already.
- Open the file using `FileShare.None`, and for reading *and* writing
- Read the contents, and leave the file open.
- When we need to write to the registry, do so using the same filehandle (so nothing else can steal the lock from us). We may need to do a `truncate()` or equivalent first.
- Close the file as part of `Registry.Dispose()`

This should be race-condition free, and means we don't have any risk of leaving a lockfile around, nor any risk of opening a registry that really truly is locked by another process.

Alternatively, if we* do* stick with lockfiles, is there any chance we can write our process-ID to the file when we open it? That means a human can check for weird edge cases (like potentially hung and invisible) and may even allow future CKAN clients to check to see if a lockfile is stale or not.

**Comment by mgsdk** [comment]
I don't have access to my development setup at the moment, but I'll play
around with fileshare permission when I get back home at the start of next
week. Hopefully Mono has a full implementation of this.

**Comment by mgsdk** [comment]
I see my rebase magic is not up to date. This should hopefully make the case for a race-condition less likely, though acquiring the lock is still not atomic. I tested the `FileShare.None` on Mono, but could easily start two instances, thus I'm still relying on creating and deleting files. Instead of doing the check manually and then creating the file, everything should be done in one go now.

Also added a special check in the GUI to show the error dialog before the main window.

**Comment by mgsdk** [comment]
The process ID is the first of the threads, though I'm not sure this is guaranteed every time the code is executed. On my system I have a total of 9 processes registered to the open CKAN instance.

**Comment by pjf** [comment]
> I see my rebase magic is not up to date. This should hopefully make the case for a race-condition less likely, though acquiring the lock is still not atomic. I tested the FileShare.None on Mono, but could easily start two instances, thus I'm still relying on creating and deleting files. Instead of doing the check manually and then creating the file, everything should be done in one go now.

Hmm, I wonder if this is a limitation of mono or something else. One of the requirements of holding a system lock on the file is that the file must remain open, so if we're doing anything to close the registry file while the process is running, then we'll also forfeit our lock on the file.

**Comment by pjf** [comment]
Oh. Locking the registry file itself is likely to cause issues with the transactional file layer¹, so maintaining a separate lockfile *does* make sense. Continuing to review, as stale locks are one of my pet subjects.

---

¹ I wish all systems used transactional filesystems, instead of no systems at all.

**Comment by pjf** [comment]
Closing this in favour of #1357, which is an extension of this work. (@mgsdk, if you're happy with that work, then do feel free to merge!)

**Pull request closed without merging**